Issue #765 Read configuration without go-bindata

kishansagathiya opened the pull request:
In the current codebase if you want to use different keys (or any
other configuration) for tests it is required to place them at a
certain path in static folder, bundle them with go-bindata and only
then run tests. This is simply unnecessary.

This commit instead directs loads configurations from file without
relying on go-bindata

Tasks:
- [x] directs loads configurations from file without
relying on go-bindata
- [x] pass the tests
- [x] since go-bindata is not being used anymore, bindata.go should be removed.

Closes #765

kishansagathiya committed:
Issue #765 Read configuration without go-bindata

In the current codebase if you want to use different keys (or any
other configuration) for tests it is required to place them at a
certain path in static folder, bundle them with go-bindata and only
then run tests. This is simply unnecessary.

This commit instead directs loads configurations from file without
relying on go-bindata

kishansagathiya committed:
Issue #765 Read configuration without go-bindata

This commit
- makes sure to use correct file path while loading them
- removes unused bindata.go

kishansagathiya committed:
Issue #765 Read configuration without go-bindata

This commit
- loads config/status-chain-genesis.json without go-bindata

kishansagathiya commented on static/static.go:
```diff
@@ -1,4 +1,2 @@
-// Package static embeds static (JS, HTML) resources right into the binaries
+// Package static contains JSON config files and keys
```
I wonder if package `static` should be renamed to something else!

adambabik replied:
Do we still need this file? There are no other Go files so I believe we can get rid of this one as well.

kishansagathiya replied:
Earlier it gave me an error something like `static does not have any go files`, but I tried just now and It worked fine even after removing this file. So, removing this in next commit.

pedropombeiro commented on params/config.go:
```diff
@@ -567,7 +566,7 @@ func (c *NodeConfig) updateGenesisConfig() error {
 // DefaultStatusChainGenesisBlock returns the StatusChain network genesis block.
 func (c *NodeConfig) DefaultStatusChainGenesisBlock() (*core.Genesis, error) {
-	genesisJSON, err := static.Asset("config/status-chain-genesis.json")
+	genesisJSON, err := ioutil.ReadFile(GetStatusHome() + "config/status-chain-genesis.json")
```
We could use `path.Join` here instead.

pedropombeiro replied:
This path doesn't seem to exist in the current PR (the config is still under `static` folder).

pedropombeiro commented on params/config.go:
```diff
@@ -715,3 +714,12 @@ func (c *NodeConfig) FormatAPIModules() []string {
+	if gopath == "" {
+		gopath = build.Default.GOPATH
+	}
+	return gopath + "/src/github.com/status-im/status-go/"
```
Same here.

pedropombeiro reviewed (changes requested):
There are missing usages of static assets to be replaced in this PR. All related code could also be removed, such as `make generate` target.

kishansagathiya committed:
Issue #765 Read configuration without go-bindata

This commit
- Fixes invalid paths
- Uses path.Join instead of + operator for joining paths

kishansagathiya committed:
Issue #765 Read configuration without go-bindata

This commit remove go-bindata related code

adambabik commented on t/utils/utils.go:
```diff
@@ -336,10 +337,36 @@ func ImportTestAccount(keystoreDir, accountFile string) error {
 	return err
 }
+
+func gettestConfigFromFile(fileName string, config *testConfig) error {
```
`getTestConfigFromFile`?

kishansagathiya replied:
I idea was to use the variable name `testConfig` as is, but sure I can change to this.

adambabik commented on t/utils/utils.go:
```diff
@@ -336,10 +337,36 @@ func ImportTestAccount(keystoreDir, accountFile string) error {
+	return nil
+}
+
+func copyFile(sourceFile string, destinationFile string) error {
```
It's fine as files are super small but if you want to play a bit with this code, I believe it can rely exclusively on streams to copy a file instead of loading it into the memory and then dumping the memory somewhere else.

kishansagathiya replied:
replacing this with a more efficient copy function

adambabik reviewed (commented):
Looks almost perfect, thanks for a great contribution!

kishansagathiya committed:
Issue #765 Read configuration without go-bindata

This commit
- removes unused static.go file
- uses a more efficient copy function

kishansagathiya commented:
@adambabik Changes made as asked

adambabik commented on t/utils/utils.go:
```diff
@@ -336,10 +337,43 @@ func ImportTestAccount(keystoreDir, accountFile string) error {
+	if err != nil {
+		return err
+	}
+	return out.Close()
```
Shouldn't it be `return nil`? We call `defer out.Close()` a few lines earlier.

kishansagathiya replied:
And that's done too

adambabik reviewed (approved):
LGTM!

kishansagathiya committed:
Issue #765 Read configuration without go-bindata

Fixing a small error

pedropombeiro commented:
@kishansagathiya if I run `go test -v -timeout 90s github.com/status-im/status-go/account -run ^TestVerifyAccountPassword$`, I see:

```
=== RUN   TestVerifyAccountPassword
WARN [08-21|13:09:11.596] cannot copy test account PK              package=status-go/t/utils error="open /tmp/accounts700250448/test-account1-status-chain.pk: no such file or directory"
--- FAIL: TestVerifyAccountPassword (0.00s)
        Error Trace:    accounts_test.go:32
        Error:          Received unexpected error:
                        open /tmp/accounts700250448/test-account1-status-chain.pk: no such file or directory
FAIL
FAIL    github.com/status-im/status-go/account  0.042s
```

Do you see the same thing?

kishansagathiya commented:
@PombeirP I do. `make ci` doesn't show in the end that this fail. I was misled by that. Will look into this.

kishansagathiya committed:
Issue #765 Read configuration without go-bindata

Fixing a small error

kishansagathiya commented:
@PombeirP This had a silly little error. My bad src, dst order was messed up. All good now.

kishansagathiya commented:
I feel like something should be done about `make ci` though
Ideally if any test fail, in the end it should say that you have failed tests. Going manually though all test result is painful. The project that I worked earlier had such a system, not sure how.
Current tests ends like this even if there was a failed test
```
--- PASS: TestTransactorSuite (0.01s)
    --- PASS: TestTransactorSuite/TestAccountMismatch (0.00s)
    --- PASS: TestTransactorSuite/TestArgsValidation (0.00s)
    --- PASS: TestTransactorSuite/TestContractCreation (0.00s)
    --- PASS: TestTransactorSuite/TestGasValues (0.00s)
        --- PASS: TestTransactorSuite/TestGasValues/noGasDef (0.00s)
        --- PASS: TestTransactorSuite/TestGasValues/gasDefined (0.00s)
        --- PASS: TestTransactorSuite/TestGasValues/gasPriceDefined (0.00s)
        --- PASS: TestTransactorSuite/TestGasValues/nilSignTransactionSpecificArgs (0.00s)
    --- PASS: TestTransactorSuite/TestLocalNonce (0.00s)
=== RUN   TestSendTxArgsValidity
--- PASS: TestSendTxArgsValidity (0.00s)
PASS
ok  	github.com/status-im/status-go/transactions	0.066s
?   	github.com/status-im/status-go/transactions/fake	[no test files]
```

pedropombeiro commented:
Strange, if I run `make ci` against the previous commit, I get this:

```
--- PASS: TestTransactorSuite (0.01s)
    --- PASS: TestTransactorSuite/TestAccountMismatch (0.00s)
    --- PASS: TestTransactorSuite/TestArgsValidation (0.00s)
    --- PASS: TestTransactorSuite/TestContractCreation (0.01s)
    --- PASS: TestTransactorSuite/TestGasValues (0.00s)
        --- PASS: TestTransactorSuite/TestGasValues/noGasDef (0.00s)
        --- PASS: TestTransactorSuite/TestGasValues/gasDefined (0.00s)
        --- PASS: TestTransactorSuite/TestGasValues/gasPriceDefined (0.00s)
        --- PASS: TestTransactorSuite/TestGasValues/nilSignTransactionSpecificArgs (0.00s)
    --- PASS: TestTransactorSuite/TestLocalNonce (0.00s)
=== RUN   TestSendTxArgsValidity
--- PASS: TestSendTxArgsValidity (0.00s)
PASS
ok      github.com/status-im/status-go/transactions     (cached)
?       github.com/status-im/status-go/transactions/fake        [no test files]
Makefile:198: recipe for target 'test-unit' failed
make: *** [test-unit] Error 1
```

kishansagathiya commented:
@PombeirP both are same, except that I skipped `Error 1`. I thought that would be because of some error that expects a test. My bad, sorry.

pedropombeiro commented:
Merged, thanks for the contribution @kishansagathiya!
